### PR TITLE
Process Operations - osquery/core Integration

### PR DESCRIPTION
**Pull request by yying**
Replaced most instances of `pid_t` with `PlatformProcess` as necessary. 
Replaced platform-specific functions such as `getpid`, `getppid`, etc. with our equivalent process abstraction operations.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/core/init.cpp [hunk 1]
```diff
@@ -97,13 +101,20 @@ enum {
 namespace {
 extern "C" {
 static inline bool hasWorkerVariable() {
-  return (getenv("OSQUERY_WORKER") != nullptr);
+  auto value = ::osquery::getEnvVar("OSQUERY_WORKER");
+  if (value) {
```
Comment (theopolis) [inline osquery/core/init.cpp]: What does this mean? Compare to `nullptr` explicitly.
```cpp
return (value != nullptr);
```

File: osquery/core/watcher.cpp [hunk 2]
```diff
@@ -97,29 +101,30 @@ PerformanceState& Watcher::getState(const std::string& extension) {
   return instance().extension_states_[extension];
 }
 
-void Watcher::setExtension(const std::string& extension, pid_t child) {
+void Watcher::setExtension(const std::string& extension,
+                           std::shared_ptr<PlatformProcess> child) {
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Pass this by `const` reference.

File: osquery/core/watcher.cpp [hunk 3]
```diff
@@ -132,7 +137,11 @@ bool Watcher::hasManagedExtensions() {
   // Setting this counter to 0 will prevent the worker from waiting for missing
   // dependent config plugins. Otherwise, its existence, will cause a worker to
   // wait for missing plugins to broadcast from managed extensions.
-  return (getenv("OSQUERY_EXTENSIONS") != nullptr);
+  auto value = getEnvVar("OSQUERY_EXTENSIONS");
+  if (value) {
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Ambiguous comparison.

File: osquery/core/watcher.cpp [hunk 4]
```diff
@@ -339,26 +356,19 @@ void WatcherRunner::createWorker() {
     return;
   }
 
-  auto worker_pid = fork();
-  if (worker_pid < 0) {
+  std::shared_ptr<PlatformProcess> worker =
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Feel free to use `auto`.

File: osquery/core/watcher.cpp [hunk 5]
```diff
@@ -339,26 +356,19 @@ void WatcherRunner::createWorker() {
...
+      PlatformProcess::launchWorker(exec_path.string(), argv_[0]);
+  if (!worker) {
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Ambiguous comparison.

File: osquery/core/watcher.cpp [hunk 6]
```diff
@@ -382,44 +392,34 @@ bool WatcherRunner::createExtension(const std::string& extension) {
     return false;
   }
 
-  auto ext_pid = fork();
-  if (ext_pid < 0) {
+  std::shared_ptr<PlatformProcess> ext_process =
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Same as above.

File: osquery/core/watcher.cpp [hunk 7]
```diff
@@ -382,44 +392,34 @@ bool WatcherRunner::createExtension(const std::string& extension) {
...
+      PlatformProcess::launchExtension(exec_path.string(),
+                                       extension,
+                                       Flag::getValue("extensions_socket"),
+                                       Flag::getValue("extensions_timeout"),
+                                       Flag::getValue("extensions_interval"),
+                                       Flag::getValue("verbose"));
+  if (!ext_process) {
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Same as above.

File: osquery/core/watcher.h [hunk 8]
```diff
@@ -113,12 +117,13 @@ class Watcher : private boost::noncopyable {
   static void unlock() { instance().lock_.unlock(); }
 
   /// Accessor for autoloadable extension paths.
-  static const std::map<std::string, pid_t>& extensions() {
+  static const std::map<std::string, std::shared_ptr<PlatformProcess>>&
```
Comment (theopolis) [inline osquery/core/watcher.h]: Is this what `clang-format` said? Wow, this is ugly. Maybe:
```cpp
using ExtensionMap = std::map<std::string, std::shared_ptr<PlatformProcess>>;
```

Can help here.

File: osquery/core/watcher.h [hunk 9]
```diff
@@ -279,16 +287,18 @@ class WatcherRunner : public InternalRunnable {
 /// The WatcherWatcher is spawned within the worker and watches the watcher.
 class WatcherWatcherRunner : public InternalRunnable {
  public:
-  explicit WatcherWatcherRunner(pid_t watcher) : watcher_(watcher) {}
+  explicit WatcherWatcherRunner(std::shared_ptr<PlatformProcess> watcher)
```
Comment (theopolis) [inline osquery/core/watcher.h]: This should take a `const` reference.

File: osquery/core/watcher.h [hunk 10]
```diff
@@ -127,20 +132,23 @@ class Watcher : private boost::noncopyable {
...
   /// Setter for worker process.
-  static void setWorker(pid_t child) { instance().worker_ = child; }
+  static void setWorker(std::shared_ptr<PlatformProcess> child) {
+    instance().worker_ = child;
+  }
 
   /// Setter for an extension process.
-  static void setExtension(const std::string& extension, pid_t child);
+  static void setExtension(const std::string& extension,
+                           std::shared_ptr<PlatformProcess> child);
```
Comment (theopolis) [inline osquery/core/watcher.h]: `const` reference for the `std::shared_ptr` container.

File: osquery/core/init.cpp [hunk 11]
```diff
@@ -531,12 +554,14 @@ void Initializer::start() const {
   }
   initLogger(binary_);
 
+#ifndef WIN32
```
Comment (theopolis) [inline osquery/core/init.cpp]: Is this gate related to the process abstractions? Or just helpful since the distributed APIs are another can of porting worms?

File: osquery/core/watcher.h (hunk 8 above)
Comment (yying) [inline osquery/core/watcher.h]: This should be post `make format`, so yes, `clang-format` appears to have not done well here...

File: osquery/core/init.cpp (hunk 11 above)
Comment (yying) [inline osquery/core/init.cpp]: I think this might be more in-tune with the later milestone of making the daemon exist as a Windows service. We can remove this for now (and another gate near the `signal` invocations in the same file) since it shouldn't affect POSIX builds.

File: osquery/core/watcher.h [hunk 12]
```diff
@@ -35,8 +39,8 @@ class WatcherRunner;
 /**
  * @brief Categories of process performance limitations.
  *
- * Performance limits are applied by a watcher thread on autoloaded extensions
  * and a optional daemon worker process. The performance types are identified
+ * Performance limits are applied by a watcher thread on autoloaded extensions
```
Comment (PoppySeedPlehzr) [inline osquery/core/watcher.h]: Did this comment line get accidentally reordered?

File: osquery/core/init.cpp (hunk 1 above)
Comment (yying) [inline osquery/core/init.cpp]: Will that work for `boost::optional`? Would explicitly casting it to a `bool` make it more clear as to the intention of the code?

Comment (PoppySeedPlehzr) [inline osquery/core/init.cpp]: It should work fine right? This seems to be how they use it in the example in their doc http://www.boost.org/doc/libs/1_60_0/libs/optional/doc/html/index.html

File: osquery/core/watcher.h [hunk 13]
```diff
@@ -127,20 +132,23 @@ class Watcher : private boost::noncopyable {
...
+  static PerformanceState& getState(const PlatformProcess& child);
   static PerformanceState& getState(const std::string& extension);
 
   /// Accessor for the worker process.
-  static pid_t getWorker() { return instance().worker_; }
+  static PlatformProcess& getWorker() { return *instance().worker_; }
 
   /// Setter for worker process.
-  static void setWorker(pid_t child) { instance().worker_ = child; }
+  static void setWorker(std::shared_ptr<PlatformProcess> child) {
```
Comment (PoppySeedPlehzr) [inline osquery/core/watcher.h]: should this also be a `const` ref?

File: osquery/core/watcher.cpp [hunk 14]
```diff
@@ -339,26 +359,18 @@ void WatcherRunner::createWorker() {
     return;
   }
 
-  auto worker_pid = fork();
-  if (worker_pid < 0) {
+  auto worker = PlatformProcess::launchWorker(exec_path.string(), argv_[0]);
+  if (worker.get() == nullptr) {
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Do you need to use `.get()` here? I believe a `std::shared_ptr` can compare safely with `nullptr`.

File: osquery/core/watcher.cpp [hunk 15]
```diff
@@ -382,44 +394,34 @@ bool WatcherRunner::createExtension(const std::string& extension) {
...
-  auto ext_pid = fork();
-  if (ext_pid < 0) {
+  auto ext_process =
+      PlatformProcess::launchExtension(exec_path.string(),
+                                       extension,
+                                       Flag::getValue("extensions_socket"),
+                                       Flag::getValue("extensions_timeout"),
+                                       Flag::getValue("extensions_interval"),
+                                       Flag::getValue("verbose"));
+  if (ext_process.get() == nullptr) {
```
Comment (theopolis) [inline osquery/core/watcher.cpp]: Same as above with `std::shared_ptr`.

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/2087/merge (Job: 2670) failed :-1:

**Comment by osqueryer** [comment]
Build origin/pr/2087/merge (Job: 2672) failed :-1:

**Comment by osqueryer** [comment]
Build origin/pr/2087/merge (Job: 2673) is good! :+1:

File: osquery/core/testing.h [hunk 16]
```diff
@@ -34,8 +34,9 @@ extern const char *kOsqueryTestModuleName;
 
 /// These are the expected arguments for our test worker process.
 extern const char *kExpectedWorkerArgs[];
+extern const size_t kExpectedWorkerArgsCount;
 
 /// These are the expected arguments for our test extensions process.
 extern const char *kExpectedExtensionArgs[];
+extern const size_t kExpectedExtensionArgsCount;
 }
-
```
Comment (PoppySeedPlehzr) [inline osquery/core/testing.h]: Is losing this newline gunna mess with things? I couldn't recall, but I thought we needed the trailing newline for some reason.

**Comment by osqueryer** [comment]
Build origin/pr/2087/merge (Job: 2674) is good! :+1:

File: osquery/core/process.h [hunk 17]
```diff
@@ -43,7 +44,7 @@ const PlatformPidType kInvalidPid = (PlatformPidType) - 1;
  * @brief Categories of process states adapted to be platform agnostic
  *
  * A process can have the following states. Unfortunately, because of operating
- * system differences. A generic state change is not directly translatable on
+ * system differences. A genericstate change is not directly translatable on
```
Comment (theopolis) [inline osquery/core/process.h]: a space was deleted here.

**Comment by osqueryer** [comment]
Build origin/pr/2087/merge (Job: 2675) is good! :+1:

**Comment by osqueryer** [comment]
Build origin/pr/2087/merge (Job: 2676) is good! :+1:
